README.md
# zbxmonitor

## Description
While monitoring a few hosts with Zabbix, it's easy to send alerts by e-mail or SMS. But when something goes wrong (SMTP server or GSM modem is down, etc.), you will not receive any error message.

ZbxMonitor resolves such problems by adding "second channel" from local desktop to Zabbix Server. ZbxMonitor periodically pulls data from Zabbix Server and shows its status in icon tray.

Presently zbxmonitor is tested with python 2.7 and zabbix 3.x and 4.x

## Usage
If's recommended to not launch zbxmonitor.py directly, but to create symbolic link to this file (eg. `ln -s zbxmonitor.py myzabbix.example.com.py`), then create config file `myzabbix.example.com.config`
and launch this symlink. You can run multiple instances of monitor on multiple Zabbix Servers this way.

## Requirements
All packages listed in `import` \*.py.

pip:
- pyzabbix
- plyer

### Config file
Config file must be named as \<script\>.config, eg. `zbxmonitor.py` uses `zbxmonitor.config` and must be present in script "home" directory. Config should be not readable by others/group.
Path to icons and sounds are relative to ./icons and ./sounds dirs.

Section `zbxCredenials` is described in `zbxmonitor.config.example`. Section `zbxOptions` description:
- ackOnly - default: True. Only unacknowledged messages will be shown. If False (or 0) monitor will show all (acknowledged and unacknowledged) messages.
- icon - default: host in `zbxCredenials`. You must have two files in ./icons dir: `host.example.com-ok.png` and `host.example.com-err.png` in this example.
- ignore_warn - default: False. True - SSL warnings will be suppressed (when connecting to Zabbix Server via https)
- interval - default: 30[s]. Pooling interval
- log_truncate - default: False. If True (or 1) log file in ./log will be truncated at start.
- min_severity - default: 0. Return only triggers with severity greater or equal than the given severity. **Change requires restart.** Range 0..5.
- notify - default: True. Messages will be send to notification daemon (eg. `dunst`). If False (or 0) you will be not notified, but status will be still enabled by left or right click on script icon.
- port - default: 10051
- text_mode - default: False. If True (or 0) zbxmonitor will work in text mode only. It is rather for testing purpose.
- wav - default: None. You must have two files (named aka icons) in ./sounds dir if specified.
- wav_player - default: /usr/bin/mpv. External wav player. Ignored if wav is not specified.

#### Filtering messages
re.search() is used.
- exclTg - default: empty list. If not empty `inclTg` is ignored. Example:<br>
`exclTg = [ "^SSL.*certificate", ".*overload.*" ]` - all messages contains strings in list will be suppressed.
- inclTg - default: empty list. Valid only if `exclTg` is empty. Example:<br>
`inclTg = [ ".*\.google\.com" ]` - only messages contains strings in list will be shown.

### Icon in tray
- left click shows current status of Zabbix Server.
- right click brings popup:
    - Close applet
    - Show all (acknowledged and unacknowledged) messages
    - Show unfiltered and unacknowledged messages

## Windows 7, 8, 10
### Requirements
- Python 2.7.17 (64 bit), python-2.7.17.amd64.msi
- Python 2.7 pycairo-1.10.0 (64-bit), py2cairo-1.10.0.win-amd64-py2.7.exe
- Python 2.7 pygobject-2.28.6 (64-bit), pygobject-2.28.6.win-amd64-py2.7.exe
- Python 2.7 pygtk-2.22.0 (64-bit), pygtk-2.22.0.win-amd64-py2.7.exe
Downloads from _python.org_ and _sourceforge.net_.

Entries: `wav` and `wav_player` are ignored in windows.<br>
I don't know how to daemonize script - there is no `fork` in windows. But you can launch script with `pythonw` not `python`, eg.:
```
<path_to_pythonw.exe> <path_to_script>
```
Notification in windows sometime works, but sometime doesn't. I don't know why - I'm very beginner windows user. Icon in tray works fine.


CHANGELOG
2020-05-17 Sun 14:19:02 PM
Fixed: context menu
Added: triggers min_severity (see config.example)

2019-11-13 Wed 12:54:59 PM
Added hostname to filters. Filtering now operates on string:
<hostname> + " " + <trigger_description>

2019-11-07 Thu 10:38:11 AM
Fixed erors in filters.

2019-11-06 Wed 09:29:15 AM
Linux: fixed error.
When no notification daemon zbxmonitor dropped connection to server silently.
Now writes down to log and exit.

2019-11-04 Mon 11:04:56 AM
Windows port

2019-11-01 Fri 16:25:13 PM
syslog has been replaced by logging (Win)

2019-10-31 Thu 12:52:12 PM
notify2 has been replaced by plyer (for Windows compatibility)


daemon.py
#!/usr/bin/env python

import sys
import os
import time
import atexit
import errno
import logging
from signal import SIGTERM

class Daemon:
    """
    A generic daemon class.

    Usage: subclass the Daemon class and override the run() method
    """
    def __init__(self, pidfile, stdin='/dev/null', stdout='/dev/null', stderr='/dev/null'):
        self.stdin = stdin
        self.stdout = stdout
        self.stderr = stderr
        self.pidfile = pidfile

    def daemonize(self, script_name, host):
        """
        do the UNIX double-fork magic, see Stevens' "Advanced
        Programming in the UNIX Environment" for details (ISBN 0201563177)
        http://www.erlenstar.demon.co.uk/unix/faq_2.html#SEC16
        """
        try:
            pid = os.fork()
            if pid > 0:
                # exit first parent
                sys.exit(0)
        except OSError, e:
            sys.stderr.write("fork #1 failed: %d (%s)\n" % (e.errno, e.strerror))
            sys.exit(1)

        # decouple from parent environment
        os.chdir("/")
        os.setsid()
        os.umask(0)

        # do second fork
        try:
            pid = os.fork()
            if pid > 0:
                # exit from second parent
                sys.exit(0)
        except OSError, e:
            sys.stderr.write("fork #2 failed: %d (%s)\n" % (e.errno, e.strerror))
            sys.exit(1)

        # redirect standard file descriptors
        sys.stdout.flush()
        sys.stderr.flush()
        si = file(self.stdin, 'r')
        so = file(self.stdout, 'a+')
        se = file(self.stderr, 'a+', 0)
        os.dup2(si.fileno(), sys.stdin.fileno())
        os.dup2(so.fileno(), sys.stdout.fileno())
        os.dup2(se.fileno(), sys.stderr.fileno())

        logging.info('%s: started', host)

        # write pidfile
        atexit.register(self.delpid)
        pid = str(os.getpid())
        file(self.pidfile,'w+').write("%s\n" % pid)

    def delpid(self):
        os.remove(self.pidfile)

    def start(self, script_name, host):
        """
        Start the daemon
        """
        # Check for a pidfile to see if the daemon already runs
        try:
            pf = file(self.pidfile,'r')
            pid = int(pf.read().strip())
            pf.close()
        except IOError:
            pid = None

        if pid:
            message = "pidfile %s already exist. Daemon already running?\n"
            sys.stderr.write(message % self.pidfile)
            sys.exit(1)

        # Start the daemon
        self.daemonize(script_name, host)
        self.run()

    def stop(self, host):
        """
        Stop the daemon
        """
        # Get the pid from the pidfile
        try:
            pf = file(self.pidfile,'r')
            pid = int(pf.read().strip())
            pf.close()
        except IOError:
            pid = None

        if not pid:
            message = "pidfile %s does not exist. Daemon not running?\n"
            sys.stderr.write(message % self.pidfile)
            return # not an error in a restart

        # Try killing the daemon process
        logging.info('%s: stopped', host)
        try:
            while 1:
                os.kill(pid, SIGTERM)
                time.sleep(0.1)
        except OSError, err:
            # print "***"
            # print err.errno == errno.ESRCH
            # print "***"
            # err = str(err)
            # if err.find("No such process") > 0:
            if err.errno == errno.ESRCH:
                # print self.pidfile
                if os.path.exists(self.pidfile):
                    os.remove(self.pidfile)
            else:
                print str(err)
                sys.exit(1)

    def restart(self):
        """
        Restart the daemon
        """
        self.stop()
        self.start()

    def run(self):
        """
        You should override this method when you subclass Daemon. It will be called after the process has been
        daemonized by start() or restart().
        """


dialog_nix.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# import platform
# if platform.system() == "Linux":
#     import gtk
# elif platform.system() == "Windows":
#     import gi
#     gi.require_version("Gtk", "3.0")
#     from gi.repository import Gtk as gtk
# else:
#     raise ValueError("Unknown OS")
import gtk
import sys


class MyDialog:

    def getPasswd(self, passwd, host):
        window = gtk.Window(gtk.WINDOW_TOPLEVEL)
        window.set_size_request(350, 60)
        window.set_position(gtk.WIN_POS_CENTER)
        window.set_title(host + " - ZBX password:")
        window.connect("delete_event", lambda w, e: gtk.main_quit())
        window.connect('key_press_event', self.escape)

        vbox = gtk.VBox(False, 0)
        window.add(vbox)
        vbox.show()

        entry = gtk.Entry()
        entry.set_max_length(50)
        entry.set_invisible_char("*")
        entry.set_visibility(False)
        entry.connect("activate", self.myCallback, entry, window, passwd)
        vbox.pack_start(entry, True, True, 0)
        entry.show()

        hbox = gtk.HBox(False, 0)
        vbox.add(hbox)
        hbox.show()

        button = gtk.Button(stock=gtk.STOCK_CANCEL)
        button.connect("clicked", lambda w: sys.exit(2))
        hbox.pack_start(button, True, True, 0)
        button.set_flags(gtk.CAN_DEFAULT)
        button.grab_default()
        button.show()

        button = gtk.Button(stock=gtk.STOCK_OK)
        button.connect("clicked", self.myCallback, entry, window, passwd)
        hbox.pack_start(button, True, True, 0)
        button.set_flags(gtk.CAN_DEFAULT)
        button.grab_default()
        button.show()

        window.show()

    def escape(self, widget, event):
        if gtk.gdk.keyval_name(event.keyval) == "Escape":
            #gtk.main_quit()
            #return False
            sys.exit(2)

    def myCallback(self, widget, entry, window, passwd):
        passwd[0] = entry.get_text()
        window.destroy()
        gtk.main_quit()



zbxmonitor.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import platform
if platform.system() == 'Linux':
    from daemon import Daemon
    import notify2
elif platform.system() == 'Windows':
    from plyer import notification
else:
    print "Unsupported OS"
    sys.exit(1)
from pyzabbix import *
from socket import *
from subprocess import call
import ConfigParser
import ast
import errno
import getpass
import gobject
import gtk
import logging
import os
import re
import time
import warnings


class GlobVars:
    def __init__(self, script_name):
        self.OS = platform.system()
        self.script_full_path = os.path.abspath(script_name)
        self.script_dir = os.path.dirname(self.script_full_path)
        self.script_name = os.path.basename(self.script_full_path)
        self.script_short_name = os.path.splitext(self.script_full_path)[0]

        self.config_file = self.script_short_name + ".config"
        self.config = ConfigParser.ConfigParser()

        self.config.read(self.config_file)
        self.zbxhost = self.config.get("zbxCredentials", "host")
        self.zbxuser = self.config.get("zbxCredentials", "user")
        try:
            self.zbxpasswd = self.config.get("zbxCredentials", "passwd")
        except:
            self.zbxpasswd = None
        try:
            self.zbxurl = self.config.get("zbxCredentials", "url")
        except:
            self.zbxurl = "https://" + self.zbxhost + "/zabbix"
        #
        self.defaults = {
            "interval": 30 * 1000,
            "notify": True,
            "port": 10051,
            "text_mode": False,
            "log_truncate": False,
            "ignore_warn": False,
            "icon": self.zbxhost,
            "wav": None,
            "wav_player": "/usr/bin/mpv",
            "ackOnly": True,
            "min_severity": 0,
            "exclTg": [],
            "inclTg": []
        }
        #
        try:
            self.log_truncate = ast.literal_eval(self.config.get("zbxOptions", "log_truncate"))
        except:
            self.log_truncate = self.defaults["log_truncate"]
        #
        self.make_dir(self.script_dir + "/log")
        flog = self.script_dir + "/log/" + self.zbxhost + ".log"
        fmode = ('w' if self.log_truncate else 'a')
        logging.basicConfig(format='%(asctime)s|%(levelname)s|%(name)s|%(message)s', filename=flog, filemode=fmode, level=logging.INFO)
        #
        self.make_dir(self.script_dir + "/tmp")
        #
        try:
            self.zbxinterval = int(self.config.get("zbxOptions", "interval")) * 1000
        except:
            self.zbxinterval = self.defaults["interval"]
        try:
            self.zbxnotify = ast.literal_eval(self.config.get("zbxOptions", "notify"))
        except:
            self.zbxnotify = self.defaults["notify"]
        try:
            self.zbxport = int(self.config.get("zbxOptions", "port"))
        except:
            self.zbxport = self.defaults["port"]
        try:
            self.text_mode = ast.literal_eval(self.config.get("zbxOptions", "text_mode"))
        except:
            self.text_mode = self.defaults["text_mode"]
        try:
            self.zbxignore_warn = ast.literal_eval(self.config.get("zbxOptions", "ignore_warn"))
        except:
            self.zbxignore_warn = self.defaults["ignore_warn"]
        try:
            self.zbxicon = self.script_dir + "/icons/" + self.config.get("zbxOptions", "icon")
        except:
            self.zbxicon = self.script_dir + "/icons/" + self.defaults["icon"]
        try:
            self.zbxwav_player = self.config.get("zbxOptions", "wav_player")
        except:
            self.zbxwav_player = self.defaults["wav_player"]
        try:
            self.zbxwav = self.script_dir + "/sounds/" + self.config.get("zbxOptions", "wav")
        except:
            self.zbxwav = self.defaults["wav"]
        try:
            self.zbxackOnly = ast.literal_eval(self.config.get("zbxOptions", "ackOnly"))
        except:
            self.zbxackOnly = self.defaults["ackOnly"]
        try:
            self.zbxmin_severity = int(self.config.get("zbxOptions", "min_severity"))
        except:
            self.zbxmin_severity = self.defaults["min_severity"]
        try:
            self.zbxExclTg = ast.literal_eval(self.config.get("zbxOptions", "exclTg"))
        except:
            self.zbxExclTg = self.defaults["exclTg"]
        if len(self.zbxExclTg) == 0:
            try:
                self.zbxInclTg = ast.literal_eval(self.config.get("zbxOptions", "inclTg"))
            except:
                self.zbxInclTg = self.defaults["inclTg"]
        else:
            self.zbxInclTg = self.defaults["inclTg"]

        self.zbx_ping = 'ok'
        self.zbx_connected = 'ok'  # ok, port_is_down, not_logged, err_getting_data_1, err_getting_data_2
        self.zbx_status = self.zbx_last_status = 'ok'  # ok, >>current stat<<
        self.zbx_filter = (True if len(self.zbxExclTg) or len(self.zbxInclTg) else False)

        if self.zbxpasswd is None:
            if self.text_mode:
                print "Zabbix URL:", self.zbxurl
                print "Zabbix user:", self.zbxuser
                self.zbxpasswd = getpass.getpass(prompt="password: ")
            else:
                md = MyDialog()
                mypass = [""]
                md.getPasswd(mypass, self.zbxurl)
                gtk.main()
                self.zbxpasswd = mypass[0]

    def make_dir(self, name):
        if not os.path.exists(name):
            try:
                os.mkdir(name)
            except OSError as e:
                if e.errno != errno.EEXIST:
                    raise


class MyDialog:

    def getPasswd(self, passwd, host):
        window = gtk.Window(gtk.WINDOW_TOPLEVEL)
        window.set_size_request(350, 60)
        window.set_position(gtk.WIN_POS_CENTER)
        window.set_title(host + " - ZBX password:")
        window.connect("delete_event", lambda w, e: gtk.main_quit())
        window.connect('key_press_event', self.escape)

        vbox = gtk.VBox(False, 0)
        window.add(vbox)
        vbox.show()

        entry = gtk.Entry()
        entry.set_max_length(50)
        entry.set_invisible_char("*")
        entry.set_visibility(False)
        entry.connect("activate", self.myCallback, entry, window, passwd)
        vbox.pack_start(entry, True, True, 0)
        entry.show()

        hbox = gtk.HBox(False, 0)
        vbox.add(hbox)
        hbox.show()

        button = gtk.Button(stock=gtk.STOCK_CANCEL)
        button.connect("clicked", lambda w: sys.exit(2))
        hbox.pack_start(button, True, True, 0)
        button.set_flags(gtk.CAN_DEFAULT)
        button.grab_default()
        button.show()

        button = gtk.Button(stock=gtk.STOCK_OK)
        button.connect("clicked", self.myCallback, entry, window, passwd)
        hbox.pack_start(button, True, True, 0)
        button.set_flags(gtk.CAN_DEFAULT)
        button.grab_default()
        button.show()

        window.show()

    def escape(self, widget, event):
        if gtk.gdk.keyval_name(event.keyval) == "Escape":
            #gtk.main_quit()
            #return False
            sys.exit(2)

    def myCallback(self, widget, entry, window, passwd):
        passwd[0] = entry.get_text()
        window.destroy()
        gtk.main_quit()


if platform.system() == 'Linux':
    class myDaemon(Daemon):
        def run(self):
            tc.tray()


class MyGtk:
    def __init__(self):
        self.statusIcon = gtk.StatusIcon()
        self.statusIcon.connect('activate', self.show_current_stat)
        self.statusIcon.connect('popup-menu', self.on_right_click)
        zbx.status("filtered")
        if gv.zbx_status != "ok":
            self.set_icon(gv.zbxicon + "-err.png")
        else:
            self.set_icon(gv.zbxicon + "-ok.png")

    def set_icon(self, icon_file):
        self.statusIcon.set_from_file(icon_file)

    def message(self, data=None, type=gtk.BUTTONS_OK):
        msg = gtk.MessageDialog(None, gtk.DIALOG_MODAL, gtk.MESSAGE_INFO, type, data)
        rval = msg.run()
        msg.destroy()
        return rval

    def make_menu(self, event_button, event_time, data=None):
        menu = gtk.Menu()
        reconnect_item = gtk.MenuItem("Reconnect to " + gv.zbxhost)
        close_item = gtk.MenuItem("Close applet " + gv.zbxhost)
        show_unfiltered_item = gtk.MenuItem("Show unfiltered triggers " + gv.zbxhost)
        show_all_item = gtk.MenuItem("Show all triggers " + gv.zbxhost)

        if gv.zbx_connected != 'ok' and gv.zbx_ping == 'ok':
            menu.append(reconnect_item)
            reconnect_item.connect_object("activate", self.reconnect_to_zbxhost, "reconnect to zbx server")
            reconnect_item.show()

        menu.append(close_item)
        # add callback
        close_item.connect_object("activate", self.close_app, "Really close?")
        close_item.show()

        if gv.zbx_connected == 'ok' and gv.zbx_ping == "ok":
            menu.append(show_all_item)
            show_all_item.connect_object("activate", self.show_all_triggers, "show all triggers")
            show_all_item.show()
            if gv.zbx_filter:
                menu.append(show_unfiltered_item)
                show_unfiltered_item.connect_object("activate", self.show_unfiltered_triggers, "show unfiltered triggers")
                show_unfiltered_item.show()

        # Popup the menu
        menu.popup(None, None, None, event_button, event_time)

    def show_current_stat(self, event):
        zbx.status("filtered")
        if gv.zbx_status != "ok":
            self.set_icon(gv.zbxicon + "-err.png")
        else:
            self.set_icon(gv.zbxicon + "-ok.png")
        m = (" filtered triggers:\n\n" if gv.zbx_filter else " current status:\n\n")
        self.message(gv.zbxhost + m + gv.zbx_status)

    def on_right_click(self, data, event_button, event_time):
        self.make_menu(event_button, event_time)

    def reconnect_to_zbxhost(self, data=None):
        zbx.login()
        self.message('Logging to ' + gv.zbxhost + ":\n" + gv.zbx_connected)

    def show_unfiltered_triggers(self, data=None):
        zbx.status("unfiltered")
        self.message(gv.zbxhost + " unfiltered triggers:\n\n" + gv.zbx_status)

    def show_all_triggers(self, data=None):
        zbx.status("all")
        self.message(gv.zbxhost + " all triggers:\n\n" + gv.zbx_status)

    def close_app(self, data=None):
        logging.info('%s: disconnected', gv.zbxhost)
        gtk.main_quit()
        # if self.message(data, gtk.BUTTONS_OK_CANCEL) == gtk.RESPONSE_OK:
        #     gtk.main_quit()


class TrayIcon:
    def __init__(self):
        self.__gmsg = MyGtk()
        if gv.OS == "Linux":
            try:
                notify2.init(gv.script_name)
            except:
                logging.error("Can't connect to notification daemon")
                sys.exit(2)

    def check(self):
        zbx.status("filtered")
        if gv.zbx_status == "ok":
            logging.info('%s status (GUI): %s', gv.zbxhost, gv.zbx_status)
            sys.stdout.flush()
        else:
            logging.warning('%s status (GUI): %s', gv.zbxhost, gv.zbx_status)
            sys.stdout.flush()
        if gv.zbx_status != gv.zbx_last_status:
            gv.zbx_last_status = gv.zbx_status
            if gv.OS == "Linux":
                n = notify2.Notification("Zabbix: " + gv.zbxhost, gv.zbx_status)
                if gv.zbx_status == "ok":
                    n.set_urgency(1)
                else:
                    n.set_urgency(2)
                try:
                    n.show()
                except:
                    logging.error("Can't connect to notification daemon")
                    sys.exit(2)
            else:
                tmo=(10 if gv.zbx_status == "ok" else 0)
                if gv.zbxnotify:
                    try:
                        notification.notify(
                            title="Zabbix: " + gv.zbxhost,
                            message=gv.zbx_status,
                            app_name=gv.script_name,
                            app_icon="",
                            timeout=tmo,
                            ticker=gv.script_short_name,
                        )
                    except:
                        logging.error("Can't connect to notification daemon")
                        pass
            if gv.zbxwav is not None and gv.OS == "Linux":
                try:
                    f = open('/dev/null', 'w')
                    if gv.zbx_status == 'ok':
                        call([gv.zbxwav_player, gv.zbxwav + "-ok.wav"], stdout=f, stderr=f)
                    else:
                        call([gv.zbxwav_player, gv.zbxwav + "-err.wav"], stdout=f, stderr=f)
                except:
                    pass
        if gv.zbx_status == 'ok':
            self.__gmsg.set_icon(gv.zbxicon + "-ok.png")
        else:
            self.__gmsg.set_icon(gv.zbxicon + "-err.png")
        return True

    def tray(self):
        gobject.timeout_add(gv.zbxinterval, self.check)
        gtk.main()


class TrayTxt:
    def __init__(self, command):
        zbx.status("unfiltered")
        if gv.zbx_status == "ok":
            logging.info('%s status (txt): %s', gv.zbxhost, gv.zbx_status)
        else:
            logging.warning('%s status (txt): %s', gv.zbxhost, gv.zbx_status)

    def check(self):
        zbx.status("unfiltered")
        if gv.zbx_status == "ok":
            logging.info('%s status (txt): %s', gv.zbxhost, gv.zbx_status)
        else:
            logging.warning('%s status (txt): %s', gv.zbxhost, gv.zbx_status)
        if gv.zbx_status != gv.zbx_last_status:
            gv.zbx_last_status = gv.zbx_status
            if gv.zbx_status == "ok":
                logging.info('%s status (txt): %s', gv.zbxhost, gv.zbx_status)
            else:
                logging.warning('%s status (txt): %s', gv.zbxhost, gv.zbx_status)
        return gv.zbx_status

    def tray(self):
        gobject.timeout_add(gv.zbxinterval, self.check)
        gobject.MainLoop().run()


class MyZbx:
    def __init__(self):
        self.zapi = ZabbixAPI(gv.zbxurl, timeout=5)
        self.zapi.session.verify = False
        if gv.zbxignore_warn:
            warnings.filterwarnings("ignore")
        self.login()

    def pingit(self):
        sock = socket(AF_INET, SOCK_STREAM)
        sock.settimeout(2)
        try:
            sock.connect((gv.zbxhost, gv.zbxport))
            gv.zbx_ping = 'ok'
        except:
            gv.zbx_ping = 'port is down'
        finally:
            sock.close()

    def login(self):
        self.pingit()
        if gv.zbx_ping == 'ok':
            try:
                self.zapi.login(gv.zbxuser, gv.zbxpasswd)
                gv.zbx_connected = 'ok'
            except:
                gv.zbx_connected = 'not logged in'
            finally:
                print "Connect: " + gv.zbx_connected

    def status(self, mode="all"):
        self.pingit()
        if gv.zbx_ping != 'ok':
            gv.zbx_status = gv.zbx_ping
            return gv.zbx_status
        if gv.zbx_connected == 'not logged in':
            gv.zbx_status = gv.zbx_connected
            return gv.zbx_status
        gv.zbx_status = self.get_triggers(mode)
        return gv.zbx_status

    def add_to_rval(self, t, rval, extmode=False):
        # print "adding\n"
        if int(t["value"]) == 1:
            if extmode:
                rval[0] += ("{0} - {1} {2}".format(t['hosts'][0]['host'], t['description'], '(Unack)' if t['unacknowledged'] else '') + "\n\n")
                return
            if gv.zbxackOnly:
                if t['unacknowledged']:
                    rval[0] += ("{0} - {1} {2}".format(t['hosts'][0]['host'], t['description'], '(Unack)' if t['unacknowledged'] else '') + "\n\n")
            else:
                rval[0] += ("{0} - {1} {2}".format(t['hosts'][0]['host'], t['description'], '(Unack)' if t['unacknowledged'] else '') + "\n\n")
        return

    def filter(self, mode, t, rval):
        if mode   == "exclude": fltList = gv.zbxExclTg
        elif mode == "include": fltList = gv.zbxInclTg
        else: return

        checked = to_add = False
        for flt in fltList:
            # print "host/flt/description:", t['hosts'][0]['host'], flt, "|", t['description'], "|"
            if re.search(flt, t['hosts'][0]['host'] + " " + t['description']):
                if mode == "exclude":
                    checked = True
                    to_add = False
                    continue
                else:
                    if checked: continue
                    to_add = True
            else:
                if mode == "exclude":
                    if checked: continue
                    to_add = True
                else:
                    checked = True
                    to_add = False
                    continue
        if to_add: self.add_to_rval(t, rval)


    def get_triggers(self, mode="all"):
        try:
            # Get a list of all issues (AKA tripped triggers)
            triggers = self.zapi.trigger.get(only_true=1,
                                             skipDependent=1,
                                             monitored=1,
                                             active=1,
                                             output='extend',
                                             expandDescription=1,
                                             selectHosts=['host'],
                                             min_severity = gv.zbxmin_severity,
                                             )

            # Do another query to find out which issues are Unacknowledged
            unack_triggers = self.zapi.trigger.get(only_true=1,
                                                   skipDependent=1,
                                                   monitored=1,
                                                   active=1,
                                                   output='extend',
                                                   expandDescription=1,
                                                   selectHosts=['host'],
                                                   withLastEventUnacknowledged=1,
                                                   min_severity = gv.zbxmin_severity,
                                                   )
            gv.zbx_connected = 'ok'
        except:
            gv.zbx_connected = "Fetch data error."
            return gv.zbx_connected

        unack_trigger_ids = [t['triggerid'] for t in unack_triggers]
        for t in triggers:
            t['unacknowledged'] = True if t['triggerid'] in unack_trigger_ids else False

        # Print a list containing only "tripped" triggers
        triggers.sort()
        rval = ['']
        for t in triggers:
            # print "description/ack:", t['description'], "|", t['unacknowledged']
            if mode == "filtered":
                if len(gv.zbxExclTg) > 0:
                    self.filter("exclude", t, rval)
                elif len(gv.zbxInclTg) > 0:
                    self.filter("include", t, rval)
                else:
                    self.add_to_rval(t, rval)
            elif mode == "unfiltered":
                self.add_to_rval(t, rval)
            elif mode == "all":
                self.add_to_rval(t, rval, True)
            else:
                return "Unknown mode: " + mode

        if rval[0] == '':
            return 'ok'
        else:
            return rval[0]


def main(argv):
    global gv, zbx, tc

    if len(argv) == 1:
        command = "start"
    else:
        command = argv[1]

    if command in ("start", "stop"):
        pass
    else:
        print "Unknown command"
        sys.exit(2)

    gv = GlobVars(argv[0])
    print gv.script_name + ":", time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(time.time())), "on", gv.zbxhost, command

    if "start" == command:
        zbx = MyZbx()
        if (gv.text_mode):
            tc = TrayTxt(command)
        else:
            tc = TrayIcon()

    if gv.OS == 'Linux':
        f = gv.script_dir + "/tmp/" + gv.zbxhost
        pidfile = f + ".pid"
        # stdoutfile = f  + ".out"
        # stderrfile = f + ".log"
        # daemon = myDaemon(pidfile, stderr=stderrfile, stdout=stdoutfile)
        daemon = myDaemon(pidfile)
        if 'start' == command:
            daemon.start(gv.script_name, gv.zbxhost)
        elif 'stop' == command:
            daemon.stop(gv.zbxhost)
        else:
            print "Unknown command"
            sys.exit(2)
    else:
        tc.tray()
    sys.exit(0)

if __name__ == '__main__':
    main(sys.argv)
